api/src/main/java/jakarta/data/repository/Streamable.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2022 Contributors to the Eclipse Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package jakarta.data.repository;
+
+import java.util.List;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+/**
+ * Simple interface to ease streamability of {@link Iterable}s.
+ * This is interface and can therefore be used as the assignment target for a lambda expression or method reference.
+ */
+@FunctionalInterface
+public interface Streamable<T> extends Iterable<T>, Supplier<Stream<T>> {
```

Comment:
Is there an intent to use this beyond pagination and as its own return type for repository methods?
`Streamable<Person> streamable = people.findByFirstName(name)`
If so, it seems like each call to streamable.get() would need to fetch the data again if allowing it to be backed by a JPA provider's `query.getResultStream()`, and then we should also be clear about requiring the user to ensure the stream is closed when done with it.  This would be fine if that's the case, but we need to spell out the requirements then so that the user can know what to rely upon from the implementation.
If it's only for pagination and restricted to operating an already-loaded list of entities for a page then we don't need to be concerned with these things.

Reply:
In Spring Data, we don't assume such indirection. Executing a query method results in the query result computed unless explicitly specified (e.g. when using `Stream`) directly. That's more of a specification question for query methods rather than a design question of the type here. Theoretically, you could also implement a query method returning an `Iterable` in a way that wraps some database cursor, but I've just never seen anyone do that because of all the nasty side effects that keeping resources open beyond the invocation of a method brings.

Reply:
How about other frameworks, e.g. Micronaut, @graemerocher ?

Reply:
We don't currently have a `Streamable` interface as of today. Could be nice though. We do allow returning `Stream` instances from finder methods though.

Reply:
Just to say from my perspective I don't see any issue with supporting this though, it seems useful